File: src/main/java/Decorator/Adapter.java
package Decorator;

class Client {
    public void test() {
        useTarget(new Decorator.ServiceAdapter());
    }

    private void useTarget(ITarget target) {
        target.operation();
    }
}

interface ITarget {
    String operation();
}

class ServiceAdapter implements ITarget { //Wrapper
    private Service service = new Service();

    @Override
    public String operation() {
        return service.complicatedLongOperationWhichShouldBeWrapped();
    }
}

class Service {
    public String complicatedLongOperationWhichShouldBeWrapped() {
        return "this is a complicatedLongOperationWhichShouldBeWrapped";
    }
}

File: src/main/kotlin/Adapter.kt
internal interface IService { // target signature
    fun operation(): String
}

internal class Service { // Service to wrap
    fun complicatedLongOperationWhichShouldBeWrapped(): String {
        return "this is a complicatedLongOperationWhichShouldBeWrapped"
    }
}

class ServiceAdapter : IService { //Wrapper
    private val service = Service()

    override fun operation(): String {
        return service.complicatedLongOperationWhichShouldBeWrapped()
    }
}

